packages/core/test/encodeFunctionCall.ts

```diff
@@ -0,0 +1,754 @@
+    });
+  });
+
+  describe("passing id", () => {
+    it("should be able to call the same function twice by passing an id", () => {
+      const moduleWithSameCallTwice = buildModule("Module1", (m) => {
+        const sameContract1 = m.contract("Example");
+
+        m.encodeFunctionCall(sameContract1, "test", [], { id: "first" });
+        m.encodeFunctionCall(sameContract1, "test", [], { id: "second" });
+
+        return { sameContract1 };
+      });
+
+      assert.equal(moduleWithSameCallTwice.id, "Module1");
+
+      const callFuture = [...moduleWithSameCallTwice.futures].find(
+        ({ id }) => id === "Module1#first"
+      );
+
+      const callFuture2 = [...moduleWithSameCallTwice.futures].find(
+        ({ id }) => id === "Module1#second"
+      );
+
+      assert.isDefined(callFuture);
+      assert.isDefined(callFuture2);
+    });
+
+    it("should throw if the same function is called twice without differentiating ids", () => {
+      assert.throws(
+        () =>
+          buildModule("Module1", (m) => {
+            const sameContract1 = m.contract("SameContract");
+            m.encodeFunctionCall(sameContract1, "test");
+            m.encodeFunctionCall(sameContract1, "test");
+
+            return { sameContract1 };
+          }),
+        `The autogenerated future id ("Module1#SameContract.test") is already used. Please provide a unique id, as shown below:
+
+m.encodeFunctionCall(..., { id: "MyUniqueId"})`
+      );
+    });
+
+    it("should throw if a call tries to pass the same id twice", () => {
+      assert.throws(
+        () =>
+          buildModule("Module1", (m) => {
+            const sameContract1 = m.contract("SameContract");
+            m.encodeFunctionCall(sameContract1, "test", [], { id: "first" });
+            m.encodeFunctionCall(sameContract1, "test", [], { id: "first" });
+            return { sameContract1 };
+          }),
+        'The future id "first" is already used, please provide a different one.'
+      );
+    });
+  });
+
+  describe("validation", () => {
+    describe("module stage", () => {
+      it("should not validate a non-contract", () => {
+        assert.throws(
+          () =>
+            buildModule("Module1", (m) => {
+              const another = m.contract("Another", []);
+              const call = m.call(another, "test");
+
+              m.encodeFunctionCall(call as any, "test");
+
+              return { another };
+            }),
+          /Invalid contract given/
+        );
+      });
+
+      it("should not validate a library", () => {
+        assert.throws(
+          () =>
+            buildModule("Module1", (m) => {
+              const another = m.library("Another");
+
+              m.encodeFunctionCall(another as any, "test");
+
+              return { another };
+            }),
+          /Invalid contract given/
+        );
+      });
+    });
+
+    it("should not validate a non-existant hardhat contract", async () => {
```

Comment:
Is this not testing the `m.contract` validation?

Reply:
No, this validation (as well as the normal function call) has its own version of the check that this is testing. 

That being said, this comment made me notice that I somehow forgot to swap the validation functions when I copied these tests from the normal function call tests, so I'll push a commit with that change.